consumer/consumer.go

```diff
@@ -170,43 +187,59 @@ func (c *Consumer) handleMessages(ch <-chan *sarama.ConsumerMessage, offset offs
 	c.Logger.Println("partition messages - reading" + logSuffix)
 
 	for msg := range ch {
-		var attempts int
-
-		for {
-			attempts++
-
-			// Note: The second returned value is not checked because we will never receive messages for a topic
-			// that it does not have a handler for.
-			h, _ := c.handlers.Get(msg.Topic)
-
-			// if an error occurs during an unformat call, it usually means
-			// the chosen converter was the wrong one. the consumption must block
-			// and an error must be reported.
-			m, err := h.Converter.FromKafka(msg)
-			if err == nil {
-				err = h.Handler.HandleMessage(m)
-				if err == nil {
-					offset.MarkOffset(msg, "")
-					if c.Metrics != nil {
-						c.Metrics.Report(*m, &Metrics{
-							Attempts:        attempts,
-							RemainingOffset: max.HighWaterMarkOffset() - msg.Offset,
-						})
-					}
-					break
-				}
-			}
-
-			c.Logger.Printf("an error occured while consuming a message: %v, retrying after %s.", err, c.config.RetryInterval)
+		m, attempts := c.handleMsg(msg)
+		if m == nil {
+			c.Logger.Println("partition messages - closing" + logSuffix)
+			break
+		}
+		// We succeeded in sending the message.
+		offset.MarkOffset(msg, "")
+		if c.Metrics != nil {
+			c.Metrics.Report(*m, &Metrics{
+				Attempts:        attempts,
+				RemainingOffset: max.HighWaterMarkOffset() - msg.Offset,
+			})
+		}
+	}
+}
 
-			select {
-			case <-time.After(c.config.RetryInterval):
-			case <-c.quit:
-				c.Logger.Println("partition messages - closing" + logSuffix)
-				return
-			}
+// handleMsg attempts to handle the message. It retries until the
+// message can be handled OK or the consumer is stopped.
+// It returns the message that was handled and the
+// number of attempts that it took, or
+// nil if the consumer was stopped.
+func (c *Consumer) handleMsg(msg *sarama.ConsumerMessage) (*Message, int) {
+	attempts := 0
+	for a := retry.StartWithCancel(c.retryStrategy, nil, c.quit); a.Next(); {
+		if !a.More() {
+			// Note: we're assuming here that the retry
+			// strategy does not terminate except when
+			// explicitly cancelled.
+			return nil, 0
+		}
+		attempts++
+
+		// Note: The second returned value is not checked
+		// because we will never receive messages for a topic
+		// that it does not have a handler for.
+		h, _ := c.handlers.Get(msg.Topic)
+
+		// If an error occurs during an unformat call, it
+		// usually means the chosen converter was the wrong one.
+		// The consumption must block and an error must be
+		// reported.
+		m, err := h.Converter.FromKafka(msg)
+		if err != nil {
+			c.Logger.Printf("an error occured while converting a message: %v; retrying", err)
+			continue
+		}
+		if err := h.Handler.HandleMessage(m); err != nil {
+			c.Logger.Printf("an error occured while handling a message: %v; retrying", err)
+			continue
 		}
+		return m, attempts
 	}
+	panic("unreachable")
```

Comment:
Is the `panic` here to avoid returning `nil, 0` and avoid the error from the linter?

Reply:
yup

Reply:
TIL thanks!